compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -855,7 +861,18 @@ class Typer extends Namer
   }
 
   def typedNamedArg(tree: untpd.NamedArg, pt: Type)(using Context): NamedArg = {
-    val arg1 = typed(tree.arg, pt)
+    val fallback: Option[Context ?=> Tree] = if (ctx.property(JavaAnnotationArg).isDefined) {
+      pt match {
+        case AppliedType(a, typ :: Nil) if (a.isRef(defn.ArrayClass)) =>
+          Some { typed(untpd.JavaSeqLiteral(tree.arg :: Nil, TypeTree(typ)), pt) }
+        case _ => None
+      }
+    } else None
+
+    val arg1 = fallback match {
+      case Some(f) => tryAlternatively { typed(tree.arg, pt) } { f }
+      case _ => typed(tree.arg, pt)
+    }
```

Comment:
I've done it that way because I wanted to avoid repeating `typed(tree.arg, pt)` three times, as it may be a source of errors when the logic would be changed. On the other hand, merging those two patterns is much more readable.